engine/src/BehaviourSystem.cpp

```diff
@@ -0,0 +1,98 @@
+#include "Engine/Intermediate/BehaviourSystem.hpp"
+
+#include "Engine/Core/Tools/BranchPrediction.hpp"
+#include "Engine/Intermediate/BehaviourComponent.hpp"
+
+using namespace GPE;
+
+BehaviourSystem* BehaviourSystem::m_pInstance{nullptr};
+
+void BehaviourSystem::addUpdate(BehaviourComponent* updateFunction) noexcept
+{
+    m_updateFunctions.emplace_back(updateFunction);
+}
+
+void BehaviourSystem::addFixedUpdate(BehaviourComponent* fixedUpdateFunction) noexcept
+{
+    m_updateFunctions.emplace_back(fixedUpdateFunction);
+}
+
+void BehaviourSystem::removeUpdate(BehaviourComponent* updateFunctionToRemove) noexcept
+{
+    for (auto&& function : m_updateFunctions)
+    {
+        if (unlikely(function == updateFunctionToRemove))
+        {
+            std::swap(function, m_updateFunctions.back());
+            m_updateFunctions.pop_back();
+        }
+    }
+}
+
+void BehaviourSystem::removeFixedUpdate(BehaviourComponent* fixedUpdateFunctionToRemove) noexcept
+{
+    for (auto&& function : m_updateFunctions)
+    {
+        if (unlikely(function == fixedUpdateFunctionToRemove))
+        {
+            std::swap(function, m_updateFunctions.back());
+            m_updateFunctions.pop_back();
+        }
+    }
+}
+
+void BehaviourSystem::addBehaviour(BehaviourComponent* pBehaviour) noexcept
+{
+    m_pBehaviours.push_back(pBehaviour);
+    pBehaviour->awake();
+}
+
+void BehaviourSystem::updateBehaviourPointer(BehaviourComponent* newPointorBehaviour,
+                                             BehaviourComponent* exPointorBehaviour) noexcept
+{
+    for (std::vector<BehaviourComponent*>::iterator it = m_pBehaviours.begin(); it != m_pBehaviours.end(); it++)
```

Comment:
```suggestion
    const std::vector<BehaviourComponent*>::iterator end = m_pBehaviours.end();
    for (std::vector<BehaviourComponent*>::iterator it = m_pBehaviours.begin(); it != end; ++it)
```

Since the size and location of the data are not changed, `m_pBehaviours.end()` will remain constant